File: tinysat-python/tinysat.py
import sys
import time

from enum import Enum

import base.dimacs_parser as parser
from base.instance import Instance
from solvers.dpll_solver import DPLLSolver


class Satisfiability(Enum):
    SAT = 'SATISFIABLE'
    UNSAT = 'UNSATISFIABLE'
    # UNKNOWN = 'UNKNOWN'


def run_solver(program):
    instance = parser.parse_program(program)
    solver = DPLLSolver("input_order")
    is_sat = solver.solve(instance)

    if is_sat:
        assignments = solver.get_assignments()
        assignments = [parser.decode_assignment(a) for a in assignments]
        return Satisfiability.SAT, assignments
    else:
        return Satisfiability.UNSAT, None


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("There are not enough arguments!")
        print("Expected: python tinysat.py <input file>")
        sys.exit()

    with open(sys.argv[1], 'r') as f:
        program = f.read()

    start_time = time.time()
    result, assignments = run_solver(program)
    elapsed_time = time.time() - start_time


    print()
    if result == Satisfiability.UNSAT:
        print("UNSAT")
    else:
        print("SAT")
        [print(a) for a in assignments]
    
    print(f"Time elapse: {elapsed_time * 1000 :.2f}ms")
    

File: tinysat-java/src/tinysat/TinySAT.java
package tinysat;

import tinysat.base.DimacsParser;
import tinysat.base.Instance;
import tinysat.base.Variable;
import tinysat.solvers.DPLLSolver;
import tinysat.solvers.Solver;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class TinySAT {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Missing arguments. Expected: java -jar TinySAT.jar <inputfile>");
            return;
        }
        String filename = args[0];
//        String filename = "../examples/simple2.dimacs";

        String program;
        try {
            program = new String(Files.readAllBytes(Paths.get(filename)));
        } catch (IOException e) {
            e.printStackTrace();
            return;
        }

        Instance ins = DimacsParser.parseProgram(program);

        Solver solver = new DPLLSolver();
        solver.solve(ins);

        List<List<Variable>> assignments = solver.getAssignments();

        if (assignments.size() == 0) {
            System.out.println("UNSATISFIABLE");
        } else {
            System.out.println("SATISFIABLE");
            for (List<Variable> assignment : assignments) {
                int[] result = DimacsParser.decodeAssignments(assignment);
                System.out.println(Arrays.toString(result));
            }

        }
    }
}
